WFLY-10784 not fail WeldStartCompletionService on optional components

kudrevatykh opened the pull request:
Thanks for submitting your Pull Request!

Please make sure your PR meets the following requirements:
- [x ] Pull Request title is properly formatted: `[WFLY-XYZ] Subject` or `WFLY-XYZ Subject`
- [ x] Pull Request contains link to the JIRA issue(s)
- [ x] Pull Request contains description of the issue(s)
- [ x] Pull Request does not include fixes for issues other than the main ticket
- [ x] Attached commits represent units of work and are properly formatted

For bigger changes, major and minor component upgrades make sure your PR also meets following requirements:
- [ ] Pull Request requires a change to the documentation
- [ ] Documentation have been updated accordingly
- [ ] Tests were added to cover changes

For new features ensure as well:
- [ ] Analysis was done
- [ ] Test Plan has been done
- [ ] Tests were verified in advance

If you are not an active contributor of the WildFly project you can request sponsorship by one of the members to help guide you through the process.

kudrevatykh committed:
WFLY-10784 fix

Not fail WeldStartCompletionService when optional components are not started

manovotn commented on weld/subsystem/src/main/java/org/jboss/as/weld/deployment/processors/WeldDeploymentProcessor.java:
```diff
@@ -293,18 +294,20 @@ public void deploy(DeploymentPhaseContext phaseContext) throws DeploymentUnitPro
+        getComponentStartServiceNames(deploymentUnit).entrySet().forEach(e->weldStartCompletionServiceBuilder.addDependencies(e.getKey(), e.getValue()));
         for (DeploymentUnit sub : subDeployments) {
-            weldStartCompletionServiceBuilder.addDependencies(getComponentStartServiceNames(sub));
+            getComponentStartServiceNames(sub).entrySet().forEach(e->weldStartCompletionServiceBuilder.addDependencies(e.getKey(), e.getValue()));
```
`ServiceBuilder.addDependencies()` is deprecated and even says that optional dependencies are unsafe and should not be used.

kudrevatykh replied:
Can you provide recipe how to depend on component when it can not start in any moment?

manovotn replied:
I don't know if it is possible at all. With current setup, Weld can do some cleanup operation allowing us to lower the footprint. If we cannot rely on all components started, then we cannot do any cleanup. It also seems that `org.jboss.msc.service.ServiceTarget` doesn't support your desired approach at all.

Can you please provide a reproducer for your problem (via JIRA - https://issues.jboss.org/browse/WFLY-10784)? If we had that, we could start fiddling with it and see what we can do.

kudrevatykh replied:
I added commit with reproducer, but can't run any test in wildlfy source tree, only in separate project

manovotn commented on weld/subsystem/src/main/java/org/jboss/as/weld/deployment/processors/WeldDeploymentProcessor.java:
```diff
@@ -321,9 +324,9 @@ public void deploy(DeploymentPhaseContext phaseContext) throws DeploymentUnitPro
-            .add(ConfigurationKey.NON_PORTABLE_MODE.get(), nonPortableMode)
-            .add(ConfigurationKey.ALLOW_OPTIMIZED_CLEANUP.get(), true)
-            .build();
+                .add(ConfigurationKey.NON_PORTABLE_MODE.get(), nonPortableMode)
```
Unnecessary formatting.

kudrevatykh replied:
fixed

manovotn reviewed (changes requested):
@kudrevatykh first and foremost, I don't understand what is the case where current solution fails (even after reading the [JIRA issue](https://issues.jboss.org/browse/WFLY-10784)). Can you provide a reproducer, detailed description, or, better still, add an automated test?

Secondly, the PR uses deprecated methods from `ServiceBuilder` which rely on `DependencyType` and according to javadoc - "_Optional dependencies are unsafe and should not be used._". So this isn't a way to go.

We first need to see what's the actual problem and have a way of reproducing it, then we can jump to solutions. That being said, we will likely need to update `WeldDeploymentProcessor` anyway to remove usage of deprecated classes.

kabir commented:
@kudrevatykh Adding the fixme label until you have addressed @manovotn 's concerns. @mention me once it is all in order

kudrevatykh committed:
reproducer for WFLY-10784

EJB service and WAR can be deployed when then independent modules, but in EAR can't be deployed

manovotn commented:
@kabir please close this PR. I will raise a different one with the approach Brian suggested on JIRA.
